### ISSUE TITLE
Automated integration tests

### ISSUE THREAD

**Issue by TheMarex** [valhalla/valhalla#1303]
Currently integration tests need to be run manually and are not enforced using the CI system. This is a major risk since changes can easily introduce regressions without anyone noticing.

The current integration test system requires you to do the following steps:

- Run the test routes on a `good` version (e.g. `master`)
- Save the results
- Run the tests routes on a `new` version (e.g. your branch)
- Manually verify the routes

Manually verifying this is error prone and time intensive. A common approach to solve this problem is to have a set of `fixtures` checked in the version control system. The easiest version of this is to have all JSON responses of all test routes committed to git and use a tool that can diff JSON fixtures in a reliable way.

Of course, the better option to do this is to use a framework like cucumber to explicitly capture what behavior should be tested with a given fixture (see #1285).

@dgearhart you were pointed out as the person that knows this system best, how would you approach automating it?

**Comment by dgearhart** [comment]
Thank you @TheMarex for submitting this proposal.

Some background…
[Quality Route Guidance](https://2015.stateofthemap.us/quality-route-guidance) has been a goal for Valhalla from the beginning. RAD (Real world Analyze Delta) testing has been a key tool to tweak route algorithms and prevent bugs from going into production.  Yes, it is time intensive - however, a route guidance developer needs to understand the problem space from data to algorithms to user - RAD testing helps the developer to acquire that knowledge and understanding. Also, items found during RAD testing should be incorporated into the automated integration testing, if possible. Automated integration tests can lead to a false sense of security and are time intensive to fix - however, the pros outweigh the cons.
I agree with your proposal in #1285  - `consider building this from scratch with a newer Cucumber version`

I believe we should have:
1. unit tests
2. integration tests
3. RAD tests

**Comment by dgearhart** [comment]
I would really to have tools/scripts that @danpat worked on to easily incorporate real world network snapshots into the cucumber tests

**Comment by TheMarex** [comment]
@dgearhart I fully agree with you that real world examples are an important component of testing guidance. However, missing/skipping manual verification steps are the most likely cause for introducing regressions, especially since we typically work under deadlines. Please consider that in an environment where not every contributor that makes a change to the guidance systems is an expert in the field, it is essential that behavior is documented in a way that can be automatically verified.

An intermediate step to switching to cucumber could be committing fixtures and automatically testing against those (using a JSON differ or similar). That would at least allow to track down the pull request that introduced a modification to the fixtures more easily.

**Comment by TheMarex** [comment]
I'm going ahead with automating the current setup as a stepping stone. For that we need the following components:

1. Something that can read the test route files and determine the dataset that is needed.
2. Something that downloads/extracts those datasets.
3. Something that prepared those datasets and executes the current toolchain on it.
4. Something that can compare the results against fixtures checked in git.

Looking at the current test route file the routes within each file seem to be spread out quite a bit (excluding the random routes):

![image](https://user-images.githubusercontent.com/606968/40743054-0020068c-6440-11e8-8e92-3ad94f774d81.png)

Each box in the above picture corresponds with the bounding box of one test routes file. Since we would need a big dataset for each test route file, it might make sense to determine the dataste

**Comment by dnesbitt61** [comment]
What constitutes a test failure? OSM data changes and code changes can both result in route differences from one run to another. Often these route differences are fine and do not indicate an error, just a different route selection. Tuning features like transition costs often lead to many route differences, some can be regarded as good changes leading to arguably better routes, while some changed routes may be arguably not as good. But sometimes these are judgment calls.

Testing smaller isolated behaviors like what is done in Cucumber tests seems to be less subject to change.

**Comment by TheMarex** [comment]
> What constitutes a test failure?

In the first iteration a failure will be any change in the response with regard to the `master` branch. This is primarily meant to know when something changed. A test with more semantic checks is not possible given the current test routes, since they don't document what is actually supposed to be checked.

>  Tuning features like transition costs often lead to many route differences, some can be regarded as good changes leading to arguably better routes, while some changed routes may be arguably not as good. But sometimes these are judgment calls.

Understood. The intend here is to document what impact code changes had on the response. That way the changes on test routes can be part of the actual code review and will ensure you always get a second pair of eyes in cause your own throughout investigation missed something.

This doesn't really change the current work flow for manual checks of JSON responses, except it becomes hard to forget doing those checks and the results are documented.

**Comment by TheMarex** [comment]
Investigating the individual test routes show a more interesting pattern:

![image](https://user-images.githubusercontent.com/606968/40997353-de08b3ba-68f3-11e8-98ce-ec8d0ca16a01.png)

Looks like we could get away with having a Germany/Spain/California datasets and some more smaller extracts for very short test routes. These OSM datasets will be fixed an stored somewhere on S3 probably.

The idea is to write a script that can:

- Read a test route file
- Compute the bounding box of the query
- Determine the right dataset for each query
- Prepare the OSM dataset using the valhalla toolchain (if necessary)
- Run the test route
- Save the results of each query as JSON
